Add PHPUnit tests workflow for multisite

thelovekesh opened the pull request:
## Summary

Fixes #5701 

This PR also updates deprecated `set-output` commands in GitHub Actions workflow files. See https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/

## Checklist

- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

thelovekesh committed:
Add PHP version in wp-env config

thelovekesh committed:
Add workflow to run PHP unit tests on multisite

thelovekesh committed:
Add npm script to run PHP unit tests on multisite

thelovekesh committed:
Fix yml syntax error

thelovekesh committed:
Remove skip condition from multisite unit tests workflow for testing

thelovekesh committed:
Use phpunit from vendor

thelovekesh committed:
Move tests with base dir as amp

thelovekesh committed:
Fix base dir path errors

thelovekesh committed:
Add matrix to run different testsuites

thelovekesh committed:
Fix yml syntax error

thelovekesh commented:
@westonruter I have configured workflow for both testsuites i.e. `default` and `external-http`. Some tests seem to be failing mainly from helper functions.  Should I fix them in the same PR? Can you please take a look if any test is need to be skipped?

westonruter commented:
Interesting. Yes, they should be fixed in this PR. For each failure, I'd suggest checking to see what the cause of the failure is. It may be due to a regular admin user not having all the same privileges as a superadmin on multisite. Therefore the fix may be just to check if its multisite, and then grant the user being tested with superadmin privileges.

Some tests probably also should be skipped if the functionality specifically is not available in multisite. For example, installing themes from the onboarding wizard I don't think is available on multisite.

It'll be interesting to see if this uncovers actual bugs with the code and not just incomplete tests!

thelovekesh committed:
Merge branch 'develop' into add/multisite-phpunit-workflow

thelovekesh committed:
Update test_amp_get_content_sanitizers_with_dev_mode for multisite environment

thelovekesh committed:
Update test_amp_has_paired_endpoint_go_to for multisite environment

Mock a new blog to create posts for AMP paired mode testing

thelovekesh commented:
@westonruter I believe the capability here should be `install_themes`. Thoughts?

https://github.com/ampproject/amp-wp/blob/650a4eb1a693c51a638527f7864ee17161762723/src/Admin/AfterActivationSiteScan.php#L214

thelovekesh committed:
Update test_get_amp_compatible_plugins_url to have super admin access in multisite env

thelovekesh committed:
Update test cases in ReaderThemesTest to have super admin access in multisite env

thelovekesh committed:
Update SiteHealthTest::test_register_async_test_endpoints to have super admin access in multisite env

thelovekesh committed:
Update SupportLinkTest::test_is_needed to have super admin access in multisite env

thelovekesh committed:
Update SupportScreenTest::test_is_needed to have super admin access in multisite env

thelovekesh committed:
Update UserRESTEndpointExtensionTest::test_update_review_panel_dismissed_for_template_mode to have super admin access in multisite env

thelovekesh committed:
Update ReaderThemeLoaderTest::test_filter_wp_prepare_themes_to_indicate_reader_theme

Update user privialge in multisite env and pass themes in wp_prepare_themes_for_js()

thelovekesh committed:
Update Test_Uninstall::test_remove_plugin_data to calculate correct DB queries in multisite env

thelovekesh committed:
Remove vague variable decleration

thelovekesh committed:
Remove vague inline comment

thelovekesh commented:
@westonruter Same issue as https://github.com/ampproject/amp-wp/pull/7288#issuecomment-1321774813 is also here but strangely it's not happening in multisite tests(using wp-env).

westonruter committed:
Add workaround for cdata_malformed_utf8_json test

westonruter commented:
> Same issue as [#7288 (comment)](https://github.com/ampproject/amp-wp/pull/7288#issuecomment-1321774813) is also here but strangely it's not happening in multisite tests(using wp-env).

@thelovekesh That's strange. I don't know what's going on. Perhaps just cherry-pick my commit as described in https://github.com/ampproject/amp-wp/pull/7288#issuecomment-1322614600.

westonruter commented on .github/workflows/build-test-measure.yml:
```diff
@@ -569,6 +569,61 @@ jobs:
+    name: "Unit Tests Multisite: PHP 7.4, WP Latest"
+    needs: pre-run
+    runs-on: ubuntu-latest
+    # if: needs.pre-run.outputs.changed-php-count > 0
```
Is this comment needed?

thelovekesh replied:
No, just used it for testing.

westonruter commented on tests/php/src/Admin/AfterActivationSiteScanTest.php:
```diff
@@ -276,7 +276,18 @@ function ( $caps, $cap ) {
 	 */
 	public function test_get_amp_compatible_plugins_url() {
-		wp_set_current_user( self::factory()->user->create( [ 'role' => 'administrator' ] ) );
+		if ( is_multisite() ) {
+			$user_id = self::factory()->user->create(
+				[
+					'role' => 'administrator',
+				]
+			);
+
+			grant_super_admin( $user_id );
+			wp_set_current_user( $user_id );
+		} else {
+			wp_set_current_user( self::factory()->user->create( [ 'role' => 'administrator' ] ) );
+		}
```
Since this same logic is being repeated in multiple files, I suggest factoring it out into a [shared trait](https://github.com/ampproject/amp-wp/tree/develop/tests/php/src/Helpers) or else to put in the base [`TestCase`](https://github.com/ampproject/amp-wp/blob/develop/tests/php/src/TestCase.php).

westonruter commented on tests/php/src/ReaderThemeLoaderTest.php:
```diff
@@ -151,12 +151,29 @@ public function test_filter_wp_prepare_themes_to_indicate_reader_theme() {
-		$themes = wp_prepare_themes_for_js();
+		$themes = wp_prepare_themes_for_js(
+			[
+				wp_get_theme( $active_theme_slug ),
+				wp_get_theme( $reader_theme_slug ),
```
Why were these needed to be added here explicitly I wonder?

thelovekesh replied:
On single-site installation, it returns all themes in an array but in the multisite environment, it only returns the current installed theme and a default theme.  Maybe because in respect to the current blog it can only fetch the theme that is installed?

westonruter replied:
Ah right. On a network, themes have to be network activated to be available for all sites.

westonruter commented on tests/php/test-uninstall.php:
```diff
@@ -569,7 +569,17 @@ public function test_remove_plugin_data() {
+
+		if ( is_multisite() ) {
+			/*
+			* Multisite has 2 additional queries:
+			* 1. Get sites query.
+			* 2. Switch to blog query.
+			*/
```
```suggestion
			 * Multisite has 2 additional queries:
			 * 1. Get sites query.
			 * 2. Switch to blog query.
			 */
```

westonruter commented on tests/php/test-amp-helper-functions.php:
```diff
@@ -2576,6 +2588,10 @@ public function data_amp_has_paired_endpoint() {
 	 * @param bool   $is_amp
 	 */
 	public function test_amp_has_paired_endpoint_go_to( $paired_url_structure, $suffix, $is_amp ) {
+		if ( is_multisite() ) {
+			switch_to_blog( self::factory()->blog->create() );
+		}
```
Why is this needed?

thelovekesh replied:
Just to ensure the working with subdirectory installation i.e. on `/path1/some-post/amp`. `/path2/some-post/amp`. Or we can write another test to ensure this?

westonruter replied:
Ah, ok. How about just adding a comment to indicate this is the reason for the blog switch?

thelovekesh replied:
I have added an inline comment.

westonruter commented on .github/workflows/build-test-measure.yml:
```diff
@@ -5,7 +5,7 @@ on:
       - develop
       # Include all release branches.
-      - '[0-9]+.[0-9]+'
+      - "[0-9]+.[0-9]+"
```
Use of double quotes instead of single quotes is just to improve convention?

thelovekesh replied:
Formatter did this. Will revert such changes.

thelovekesh committed:
Fix comments indentation

Co-authored-by: Weston Ruter <westonruter@google.com>

thelovekesh committed:
Add trait for mocking admin and super admin user

thelovekesh committed:
Update admin and super admin user mocking with helper MockAdminUser trait

thelovekesh committed:
Merge branch 'develop' into add/multisite-phpunit-workflow

thelovekesh committed:
Remove comment from job condition

thelovekesh committed:
Add steps to update PHPUnit version

thelovekesh committed:
Revert quote_type to single

thelovekesh committed:
Update deprecated set-output commands with GITHUB_OUTPUT environment files

westonruter commented:
> I believe the capability here should be `install_themes`. Thoughts?

@thelovekesh yes, I think you're right. The page requires `install_themes`: https://github.com/WordPress/wordpress-develop/blob/04f55a430e2ee5a88e9b5b5485b75a68fabd5a83/src/wp-admin/theme-install.php#L15-L17

westonruter commented on tests/php/src/Admin/UserRESTEndpointExtensionTest.php:
```diff
@@ -68,7 +68,18 @@ public function test_update_review_panel_dismissed_for_template_mode() {
 		$this->user_rest_endpoint_extension->register_rest_field();
 
-		$admin_user  = self::factory()->user->create_and_get( [ 'role' => 'administrator' ] );
+		$admin_user = self::factory()->user->create_and_get( [ 'role' => 'administrator' ] );
+
+		if ( is_multisite() ) {
+			$admin_user = self::factory()->user->create_and_get(
+				[
+					'role' => 'administrator',
+				]
+			);
+
+			grant_super_admin( $admin_user->ID );
+		}
+
```
This should also use `MockAdminUser`

thelovekesh replied:
@westonruter Please note the use of `create_and_get` and later this test also requires the WP user object.

westonruter replied:
You could make `MockAdminUser` return the object to use later.

thelovekesh committed:
Add inline comment for switching blogs in multisite tests

thelovekesh committed:
Update mock_admin_user() to return WP_User object

thelovekesh committed:
Update mocking admin user with mock_admin_user()

westonruter commented on tests/php/src/Helpers/MockAdminUser.php:
```diff
@@ -0,0 +1,41 @@
+		$admin_user = self::factory()->user->create_and_get( [ 'role' => 'administrator' ] );
+
+		if ( is_multisite() ) {
+			$admin_user = self::factory()->user->create_and_get(
+				[
+					'role' => 'administrator',
+				]
+			);
+
```
This user object seems to be created as a duplicate.

```suggestion
```

thelovekesh replied:
Ahh. yes.

thelovekesh committed:
Fix duplicate WP user creation

Co-authored-by: Weston Ruter <westonruter@google.com>

thelovekesh committed:
Update current user capability to install_themes

thelovekesh committed:
Update AfterActivationSiteScanTest::test_get_amp_compatible_themes_url to mock super admin user in multisite
